stock_picking_invoice_product_group/wizard/stock_invoice_onshipping.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Alex Comba - Agile Business Group
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from openerp import models, api, fields
+
+
+class StockInvoiceOnshipping(models.TransientModel):
+    _inherit = 'stock.invoice.onshipping'
+
+    group_type = fields.Selection(
+        [('group_by_product', 'Group by product'),
+         ('group_by_product_category', 'Group by product category')
+         ],
+        string='Group Type',
+    )
+
+    @api.model
+    def _group_moves_by_group_type(self, moves, group_type):
+        """Return moves grouped by group_type"""
+        grouped_moves = {}
+        for move in moves:
+            if group_type == 'group_by_product':
+                key = (move._get_partner_id(), move._get_product_id())
```

Comment:
You don't need an specific function to be called here. You can use attrgetter. Example: https://github.com/odoo/odoo/blob/8.0/addons/crm/base_partner_merge.py#L463